device/src/region/eu868/mod.rs

```diff
@@ -0,0 +1,73 @@
+#![allow(dead_code)]
+use super::*;
+use lorawan_encoding::maccommands::ChannelMask;
+
+const JOIN_CHANNELS: [u32; 3] = [868_100_000, 868_300_000, 868_500_000];
+
+#[derive(Default)]
+pub struct EU868 {
+    subband: Option<u8>,
+    last_tx: (u8, u8),
+    cf_list: Option<[u32; 5]>,
+}
+
+impl EU868 {
+    pub fn new() -> EU868 {
+        Self::default()
+    }
+}
+
+use super::JoinAccept;
+
+impl RegionHandler for EU868 {
+    fn process_join_accept<T: core::convert::AsRef<[u8]>, C>(
+        &mut self,
+        join_accept: &super::DecryptedJoinAcceptPayload<T, C>,
+    ) -> JoinAccept {
+        let mut new_cf_list = [0, 0, 0, 0, 0];
+        if let Some(cf_list) = join_accept.c_f_list() {
+            for (index, freq) in cf_list.iter().enumerate() {
+                new_cf_list[index] = freq.value();
+            }
+        }
+        self.cf_list = Some(new_cf_list);
+        JoinAccept {
+            cflist: Some(new_cf_list),
+        }
+    }
+
+    fn set_channel_mask(&mut self, _chmask: ChannelMask) {
+        // No subband in this region
+    }
+
+    fn set_subband(&mut self, subband: u8) {
+        self.subband = Some(subband);
+    }
+
+    fn get_join_frequency(&mut self, random: u8) -> u32 {
+        let channel = random as usize % JOIN_CHANNELS.len();
+        JOIN_CHANNELS[channel]
+    }
+
+    fn get_data_frequency(&mut self, random: u8) -> u32 {
+        if let Some(cf_list) = self.cf_list {
+            let channel = random as usize & 0b111;
+            if channel < JOIN_CHANNELS.len() {
+                JOIN_CHANNELS[channel]
+            } else {
+                cf_list[channel - JOIN_CHANNELS.len()]
+            }
+        } else {
+            let channel = random as usize % JOIN_CHANNELS.len();
+            JOIN_CHANNELS[channel]
+        }
+    }
+
+    fn get_join_accept_frequency1(&self) -> u32 {
+        JOIN_CHANNELS[self.last_tx.1 as usize]
```

Comment:
I did not notice where `last_tx` is set in this type. Did I miss it? :slightly_smiling_face: Could you explain what is the reason to have `last_tx` be part of the `RegionHandler`? Intuitively it feels better for me to be part of the device state somehow.

Reply:
Nope, I missed it! I put it in.

In general, I'm trying to keep anything that's considered "regional lorawan spec" inside of the region module, and anything that's considered part of the core spec in the device state machine. IMO, that should includes state as well as logic.